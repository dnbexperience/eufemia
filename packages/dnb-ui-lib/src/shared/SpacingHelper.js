/**
 * Spacing helper
 *
 */

import { createStyleObject } from '../components/space/SpacingHelper'

export const SpacingHelper = createStyleObject
