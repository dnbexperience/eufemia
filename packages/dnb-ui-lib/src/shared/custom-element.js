/**
 * Make React component aviable as a Web Component
 *
 */

import React from 'react'
import { render, unmountComponentAtNode } from 'react-dom'
import { ErrorHandler } from './error-helper'

// import "custom-element-polyfill" - insted of import 'document-register-element' // https://github.com/WebReflection/document-register-element
// This way we can controll the execution of the polyfill with customElementPolyfill()
import customElementPolyfill from './custom-element-polyfill'

export const registeredElements = (global.registeredElements =
  global.registeredElements || [])

export const registerElement = (
  tagName,
  ReactComponent,
  propNames = null,
  // remove id, because we never can have more than one of the same id
  { attributesBlacklist = ['id'] } = {}
) => {
  if (!tagName) tagName = ReactComponent.displayName || ReactComponent.name

  // stop here if we already have registered the tag
  if (registeredElements.indexOf(tagName) !== -1) return
  registeredElements.push(tagName)

  if (typeof document === 'undefined' || typeof window === 'undefined') {
    return null
  }

  //always run the customElementPolyfill unlress we are in the build process
  if (!global.registeredElements.hasPolyfill) {
    global.registeredElements.hasPolyfill = true
    customElementPolyfill(window)
  }

  if (propNames) {
    propNames = prepareDefaultProps(propNames)
  }

  class HtmlClass extends HTMLElement {
    static get observedAttributes() {
      return propNames || []
    }
    constructor(props) {
      super(props)
      this._elementRef = React.createRef()
      this._customMethodes = {}
      this._customEvents = []
      this._isConnected = false
    }
    connectedCallback() {
      this.updateChildren()
      this.renderElement()
      this._isConnected = true
    }
    attributeChangedCallback(attrName, oldAttr, newAttr) {
      if (!this._isConnected || oldAttr === newAttr) {
        return false
      }
      this.renderElement()
      return newAttr
    }
    // adoptedCallback: Invoked when the custom element is moved to a new document.
    detachedCallback() {
      unmountComponentAtNode(this)
      if (this._children) delete this._children
      if (this._isConnected) delete this._isConnected
      if (this._elementRef) delete this._elementRef
      if (this._customMethodes) delete this._customMethodes
      if (this._customEvents) delete this._customEvents
    }
    updateChildren() {
      this._children = []
      let i,
        cn = this.childNodes,
        v
      for (i = cn.length; i--; ) {
        v = toVdom(cn[i])
        if (v) {
          this._children.push(v)
        }
        // TODO: we may remove this child - need more testing
        // cn[i].remove()
      }
    }
    connectEvents(props) {
      if (props.events) {
        props.event = props.events
        delete props.events
      }

      // check if there are more than one events
      let events = props.event ? props.event.split(',') : []

      // check if there are custom renderer, if so, add them as well
      if (ReactComponent.renderProps) {
        events = Object.entries(ReactComponent.renderProps)
          .filter(([key]) => key && props[key])
          .reduce((events, [key]) => {
            events.push(key + '=' + props[key])
            delete props[key]
            return events
          }, events)
      }

      if (events.length > 0) {
        events.forEach(eventDef => {
          // extract the prop name and callback function
          let [type, func] = eventDef.split('=')
          type = EVENT_TRANSLATIONS[type] || type

          // add a react function prop or event callback
          props[type] = (...args) => {
            try {
              // check if there is a element returned, convert it to html then
              if (args[0]) {
                if (React.isValidElement(args[0])) args[0] = [args[0]]
                if (Array.isArray(args[0])) {
                  const elems = []
                  // we have to overwrite the first arg like this - and cant use map/reduce here
                  args[0].forEach(elem => {
                    if (React.isValidElement(elem)) {
                      const rootEl = document.createElement('div') // createDocumentFragment
                      render(elem, rootEl)
                      elems.push(rootEl)
                    }
                  })
                  if (elems.length > 0) args[0] = elems
                }
              }

              // call the function, either it in a class or not
              let [scope, fn] = func.split('.')
              fn = fn ? window[scope][fn] : window[scope] // TODO: remove this because of security notation
              const ret = fn.apply(scope, [...args])

              // convert to react if we get an HTMLElement
              // this is used for custom renderer
              if (ret instanceof HTMLElement) {
                const children = [],
                  cn = ret.childNodes,
                  a = ret.attributes,
                  props = {}

                for (let i = cn.length; i--; ) {
                  children.push(toVdom(cn[i])) // TODO: remove this because of security notation
                  // TODO: we may remove this child - need more testing
                  // cn[i].remove()
                }

                for (let i = a.length; i--; ) {
                  props[PROP_TRANSLATIONS[a[i].name] || a[i].name] =
                    a[i].value // TODO: remove this because of security notation
                }

                const nodeName = ret.nodeName.toLowerCase()
                ret.remove()

                return React.createElement(nodeName, props, children)
              }

              return ret
            } catch (error) {
              new ErrorHandler(
                `The '${type}' event has failed. '${func}' has to exist on a 'window' scope!`,
                error
              )
            }
          }
        })

        // do send this event to the react props
        delete props.event
      }
      return props
    }
    addEvent(eventName, eventCallback) {
      const eventWrapper = event => eventCallback.apply(this, [event])
      this._customEvents.push({ eventName, eventCallback, eventWrapper })
      return eventWrapper
    }
    removeEvent(eventId, removeCallback = null) {
      this._customEvents = this._customEvents.reduce(
        (accumulator, current) => {
          if (removeCallback) {
            const { eventCallback: eventWrapper } = current
            if (eventWrapper !== removeCallback) {
              accumulator.push(current)
            }
          } else {
            const { eventWrapper } = current
            if (eventWrapper !== eventId) {
              accumulator.push(current)
            }
          }
          return accumulator
        },
        []
      )
    }
    fireEvent(eventName, ...args) {
      this._customEvents.forEach(({ eventName: name, eventCallback }) => {
        if (name === eventName && typeof eventCallback === 'function') {
          eventCallback.apply(this, [...args])
        }
      })
    }
    renderElement() {
      let props = {},
        i = 0,
        a = this.attributes

      for (i = a.length; i--; ) {
        props[a[i].name] = a[i].value // TODO: remove this because of security notation
      }

      props = this.connectEvents(props)

      // we dont allow ids
      for (i = attributesBlacklist.length; i--; ) {
        if (props[attributesBlacklist[i]]) {
          // TODO: remove this because of security notation
          this.removeAttribute(attributesBlacklist[i]) // TODO: remove this because of security notation
        }
      }

      if (this._children && this._children.length > 0) {
        props.children = this._children // we have returned a () => before
      }

      if (this._elementRef) {
        props.ref = this._elementRef
      }
      if (!props.custom_element) {
        props.custom_element = this
      }
      if (!props.custom_method) {
        props.custom_method = (methodName, methodFunc) => {
          this[methodName] = this._customMethodes[methodName] = methodFunc
        }
      }

      render(<ReactComponent {...props} />, this)
    }
  }

  return window.customElements.define(tagName, HtmlClass)
}

// remove react props witch has uppercase chars
const filterProps = key =>
  key && !/[A-Z]/.test(key) && !/children/.test(key)

export const prepareDefaultProps = defaultProps =>
  Array.isArray(defaultProps)
    ? defaultProps.filter(filterProps)
    : Object.entries(defaultProps || {})
        .reduce((props, [key]) => {
          props.push(key)
          return props
        }, [])
        .filter(filterProps)

const toVdom = (elem, name = null) => {
  if (elem.nodeType === 3) return elem.nodeValue
  if (elem.nodeType !== 1) return null

  let children = [],
    props = {},
    i = 0,
    a = elem.attributes,
    cn = elem.childNodes

  for (i = a.length; i--; ) {
    // a[i].name = PROP_TRANSLATIONS[a[i].name]||a[i].name
    props[a[i].name] = a[i].value // TODO: remove this because of security notation
  }
  for (i = cn.length; i--; ) {
    children[i] = toVdom(cn[i]) // TODO: remove this because of security notation
  }
  props.key = `key${Math.random() * 1000}`

  return React.createElement(
    name || elem.nodeName.toLowerCase(),
    props,
    children
  )
}

const PROP_TRANSLATIONS = {
  class: 'className',
  for: 'htmlFor'
}
const EVENT_TRANSLATIONS = {
  onclick: 'onClick'
}
