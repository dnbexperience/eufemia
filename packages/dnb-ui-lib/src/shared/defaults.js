/**
 * DNB Eufemia Defaults
 *
 */

export const LOCALE = 'nb-NO'
export const CURRENCY = 'NOK'
export const CURRENCY_DISPLAY = 'symbol' // code, name, symbol
