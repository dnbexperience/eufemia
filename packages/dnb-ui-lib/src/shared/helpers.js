/**
 * Global helpers
 *
 */

import {
  warn,
  PLATFORM_MAC,
  PLATFORM_WIN,
  PLATFORM_LINUX,
  PLATFORM_IOS
} from './component-helper'

export let IS_IE11 = false
export let IS_EDGE = false
export let IS_IOS = false
export let IS_SAFARI = false
export let IS_WIN = false
export let IS_MAC = false
export let IS_LINUX = false

export const isMac = () =>
  (IS_MAC =
    typeof navigator !== 'undefined' &&
    new RegExp(PLATFORM_MAC, 'i').test(navigator?.platform))

export const isWin = () =>
  (IS_WIN =
    typeof navigator !== 'undefined' &&
    new RegExp(PLATFORM_WIN, 'i').test(navigator?.platform))

export const isLinux = () =>
  (IS_LINUX =
    typeof navigator !== 'undefined' &&
    new RegExp(PLATFORM_LINUX, 'i').test(navigator?.platform))

export const isiOS = () =>
  (IS_IOS =
    typeof navigator !== 'undefined' &&
    new RegExp(PLATFORM_IOS, 'i').test(navigator?.platform))

export const isSafari = () =>
  (IS_SAFARI =
    typeof navigator !== 'undefined' &&
    /safari/i.test(navigator?.userAgent) &&
    !/chrome/i.test(navigator?.userAgent))

export const isIE11 = () =>
  (IS_IE11 =
    typeof window !== 'undefined' && typeof document !== 'undefined'
      ? !!window.MSInputMethodContext && !!document.documentMode
      : false)

export const isEdge = () =>
  (IS_EDGE =
    typeof navigator !== 'undefined' && /edge/i.test(navigator?.userAgent))

isIE11()
isEdge()
isiOS()
isSafari()
isWin()
isMac()
isLinux()

const pageFocusElements = {}
export function setPageFocusElement(selectorOrElement, key = 'default') {
  return (pageFocusElements[key] = selectorOrElement)
}

export function applyPageFocus(key = 'default', callback = null) {
  try {
    let element = pageFocusElements[key]
    if (typeof element === 'string' && typeof document !== 'undefined') {
      element = document.querySelector(element)
    } else if (!element && typeof document !== 'undefined') {
      element = document.querySelector('.dnb-no-focus')
    }
    if (element instanceof HTMLElement) {
      if (
        [
          'h1',
          'h2',
          'h3',
          'h4',
          'h5',
          'h6',
          'p',
          'div',
          'main',
          'nav',
          'header',
          'footer',
          'aside',
          'section',
          'article'
        ].includes(String(element.nodeName).toLowerCase())
      ) {
        if (!element.hasAttribute('tabindex')) {
          element.setAttribute('tabindex', '-1')
        }
        if (
          element.classList &&
          !element.classList.contains('dnb-no-focus')
        ) {
          element.classList.add('dnb-no-focus')
        }
      }

      element.focus({ preventScroll: true })
      if (typeof callback === 'function') {
        callback(element)
      }
    }
  } catch (e) {
    warn('Error on applyPageFocus:', e)
  }
}

export function getOffsetTop(elem) {
  let offsetTop = 0
  do {
    if (!isNaN(elem.offsetTop)) {
      offsetTop += elem.offsetTop
    }
  } while ((elem = elem.offsetParent))
  return offsetTop
}

export function getOffsetLeft(elem) {
  let offsetLeft = 0
  do {
    if (!isNaN(elem.offsetLeft)) {
      offsetLeft += elem.offsetLeft
    }
  } while ((elem = elem.offsetParent))
  return offsetLeft
}

export function scrollToLocationHashId({
  offset = 0,
  delay = null,
  onCompletion = null
} = {}) {
  if (
    typeof document !== 'undefined' &&
    typeof window !== 'undefined' &&
    window.location
  ) {
    try {
      let _timeout
      const id = String(window.location.hash).replace('#', '')
      if (id.length > 0) {
        const handleScroll = () => {
          const runScroll = () => {
            const top = getOffsetTop(elem) - offset
            try {
              if (typeof IntersectionObserver !== 'undefined') {
                const intersectionObserver = new IntersectionObserver(
                  (entries) => {
                    const [entry] = entries
                    if (entry.isIntersecting) {
                      intersectionObserver.unobserve(elem)
                      if (typeof onCompletion === 'function') {
                        onCompletion(elem)
                      }
                    }
                  }
                )
                // start observing
                intersectionObserver.observe(elem)
              }

              if (window.scrollTo) {
                window.scrollTo({
                  top,
                  behavior: 'smooth'
                })
              } else {
                window.scrollTop = top
              }
            } catch (e) {
              warn('Error on scrollToLocationHashId:', e)
            }
          }

          if (delay > 0) {
            clearTimeout(_timeout)
            _timeout = setTimeout(runScroll, delay) // to make sure we run our scrollTo after the native anchor
          } else {
            runScroll()
          }
        }

        const elem = document.getElementById(id)
        if (elem instanceof HTMLElement) {
          window.addEventListener('beforeunload', () =>
            clearTimeout(_timeout)
          )
          if (document.readyState === 'loading') {
            document.addEventListener('DOMContentLoaded', handleScroll)
          } else {
            handleScroll()
          }
        }

        return elem
      }
    } catch (e) {
      warn('Error on scrollToLocationHashId:', e)
    }
  }
}

// Credit David Walsh (https://davidwalsh.name/javascript-debounce-function)

// Returns a function, that, as long as it continues to be invoked, will not
// be triggered. The function will be called after it stops being called for
// N milliseconds. If `immediate` is passed, trigger the function on the
// leading edge, instead of the trailing.
export function debounce(
  func,
  wait = 250,
  { immediate = false, context = null } = {}
) {
  let timeout
  let recall

  // This is the function that is actually executed when
  // the DOM event is triggered.
  return function executedFunction() {
    // Store the context of this and any
    // parameters passed to executedFunction
    const ctx = context || this
    const args = arguments

    if (typeof recall === 'function') {
      recall()
    }

    // The function to be called after
    // the debounce time has elapsed
    const later = () => {
      // null timeout to indicate the debounce ended
      timeout = null

      // Call function now if you did not on the leading end
      if (!immediate) {
        recall = func.apply(ctx, args)
      }
    }

    // Determine if you should call the function
    // on the leading or trail end
    const callNow = immediate && !timeout

    // This will reset the waiting every function execution.
    // This is the step that prevents the function from
    // being executed because it will never reach the
    // inside of the previous setTimeout
    clearTimeout(timeout)

    // Restart the debounce waiting period.
    // setTimeout returns a truthy value (it differs in web vs node)
    timeout = setTimeout(later, wait)

    // Call immediately if you're dong a leading
    // end execution
    if (callNow) {
      recall = func.apply(ctx, args)
    }
  }
}

export function insertElementBeforeSelection(elem) {
  try {
    const selection = window.getSelection()
    const range = selection.getRangeAt(0)
    range.cloneRange().insertNode(elem)
    selection.addRange(range) // Restore the original selection - this is a Safari fix!

    // For now I (Tobias) could not find any reason for supporting document.selection as well
    // const range = document.selection.createRange()
    // range.collapse(true)
    // range.pasteHTML(elem.outerHTML)
  } catch (e) {
    //
  }
}

export function getSelectedText() {
  try {
    return window.getSelection().toString()

    // For now I (Tobias) could not find any reason for supporting document.selection as well
    // return document.selection.createRange().text
  } catch (e) {
    //
  }
  return ''
}

export function hasSelectedText() {
  return getSelectedText().length > 0
}

export function getSelectedElement() {
  try {
    const selection = window.getSelection()
    if (selection.rangeCount > 0) {
      let elem = selection.getRangeAt(0).startContainer
      if (elem && typeof elem === 'object') {
        elem = elem.parentNode
      }
      return elem
    }

    // For now I (Tobias) could not find any reason for supporting document.selection as well
    // return document.selection.createRange().parentElement()
  } catch (e) {
    //
  }

  return null
}

export async function copyToClipboard(string) {
  if (typeof window === 'undefined' || typeof document === 'undefined') {
    return false
  }

  // get the selection range
  const selection = window.getSelection()
  const range =
    selection.rangeCount > 0 // Check if there is any content selected previously
      ? selection.getRangeAt(0) // Store selection if found
      : false // Mark as false to know no selection existed before

  const resetSelection = () => {
    try {
      // If a selection existed before copying
      selection.removeAllRanges() // Unselect everything on the HTML document
      selection.addRange(range) // Restore the original selection
    } catch (e) {
      //
    }
  }

  const copyFallback = () => {
    try {
      // create the focusable element
      const elem = document.createElement('textarea')
      elem.value = String(string)
      elem.contentEditable = true
      elem.readOnly = false
      elem.style.position = 'fixed'
      elem.style.top = '-1000px'
      document.body.appendChild(elem)

      // iOS helper
      // But right now, we do not use that
      // if (IS_IOS) {
      //   const newRange = document.createRange()
      //   newRange.selectNodeContents(elem)
      //   const sel = window.getSelection()
      //   sel.removeAllRanges()
      //   sel.addRange(newRange)
      //   elem.setSelectionRange(0, 999999)
      // } else {
      //   elem.select()
      // }

      // NB: copy only works as a result of a user action (e.g. click events)
      const success = document.execCommand('copy')

      // Cleanup
      document.body.removeChild(elem)

      resetSelection()

      if (success) {
        return true
      }
    } catch (e) {
      return e
    }

    return `Could not copy! Unknown reason. ${string}`
  }

  let success

  if (typeof navigator !== 'undefined' && navigator?.clipboard) {
    try {
      await navigator.clipboard.writeText(String(string))
      success = true
      resetSelection()
    } catch (e) {
      success = e
      const newTry = copyFallback()
      if (newTry === true) {
        success = newTry
      }
    }
  } else {
    // use the fallback as the primary, because we get
    success = copyFallback()
  }

  return success
}
