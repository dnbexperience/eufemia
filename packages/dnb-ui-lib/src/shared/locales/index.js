// import default locales
import enUS from './en-US'
import nbNO from './nb-NO'

export default { ...enUS, ...nbNO }
