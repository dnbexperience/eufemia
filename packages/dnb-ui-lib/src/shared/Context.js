/**
 * Web Context Context
 *
 */

import { createContext } from 'react'
import { LOCALE, CURRENCY, CURRENCY_DISPLAY } from './defaults'
import defaultLocales from './locales'

export const defaultContext = (props = {}) => {
  const locales = props.locales || defaultLocales
  return {
    locale: LOCALE,
    locales,
    translation: locales[props.locale || LOCALE] || locales[LOCALE] || {},
    currency: CURRENCY,
    currency_display: CURRENCY_DISPLAY,
    formRow: null,
    globalStatus: null,
    ...props
  }
}

// If no provider is given, we use the defualt context from here
const Context = createContext(defaultContext())

export default Context
