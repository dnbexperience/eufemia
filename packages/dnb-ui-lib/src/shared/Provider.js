/**
 * Lib Provider
 *
 */

import React from 'react'
import PropTypes from 'prop-types'
import Context, { defaultContext } from './Context'

// fill with data
import { prepareFormRowContext } from '../components/form-row/FormRow'

const Provider = ({ children, ...props }) => {
  // 1. Set defualt context to be overwirtter by the provider props
  const context = defaultContext(props)

  // 2. The reset will extend the Provider Context

  // chore preparation
  if (context.formRow) {
    context.formRow = prepareFormRowContext(context.formRow)
  }

  // general context update
  if (!context.update) {
    context.update = props => setContext({ ...context, ...props })
  }

  // make it posible to change the locale during runtime
  if (!context.setLocale) {
    context.setLocale = locale =>
      setContext({ ...context, translation: context.locales[locale] })
  }

  const [usedContext, setContext] = React.useState(context)

  return (
    <Context.Provider value={usedContext}>{children}</Context.Provider>
  )
}
Provider.propTypes = {
  children: PropTypes.node.isRequired
}

export default Provider
