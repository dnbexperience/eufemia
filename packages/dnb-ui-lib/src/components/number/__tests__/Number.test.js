/**
 * Component Test
 *
 */

import React from 'react'
import {
  mount,
  axeComponent,
  toJson,
  loadScss,
  mockGetSelection
} from '../../../core/jest/jestSetup'
import { LOCALE } from '../../../shared/defaults'
import { isMac } from '../../../shared/helpers'
import Component, {
  cleanNumber,
  cleanDirtyNumber,
  copyNumber,
  copySelectedNumber
} from '../Number'

const element = Component.defaultProps.element
const locale = LOCALE
const value = 12345678.9876
const snapshotProps = {
  value,
  locale,
  element
}

// make it possible to change the navigator lang
// because "navigator.language" defaults to en-US
let languageGetter, platformGetter

beforeAll(() => {
  languageGetter = jest.spyOn(window.navigator, 'language', 'get')
  platformGetter = jest.spyOn(window.navigator, 'platform', 'get')

  // simulate mac, has to run on the first render
  platformGetter.mockReturnValue('Mac')
  languageGetter.mockReturnValue(locale)

  isMac() // just to uptate the exported const: IS_MAC

  mockGetSelection()
})

describe('Node', () => {
  it('has icu and full-icu support', () => {
    expect(typeof process.versions.icu).toBe('string')

    const intl = new Intl.NumberFormat(locale, {
      style: 'currency',
      currency: 'NOK'
    })
    expect(intl.format(value)).toBe('kr 12 345 678,99') // Rounds
  })
  it('supports setting navigator.language (JSDOM)', () => {
    expect(navigator.language).toBe(locale)
  })
})

describe('Number component', () => {
  const slector = element + '.dnb-number'
  it('have to match default number snapshot', () => {
    const Comp = mount(<Component {...snapshotProps} />)
    expect(toJson(Comp)).toMatchSnapshot()
  })
  it('have to match default number', () => {
    const Comp = mount(<Component value={value} />)
    expect(Comp.find(slector).first().text()).toBe('12 345 678,9876')
  })
  it('have to match currency', () => {
    const Comp = mount(<Component value={-value} currency />)

    expect(Comp.find(slector).first().text()).toBe('kr -12 345 678,98')

    expect(
      Comp.find(slector).first().instance().getAttribute('aria-label')
    ).toBe('-12 345 678,98 norske kroner')

    // also check the formatting with one digit less
    Comp.setProps({
      children: null,
      decimals: 0,
      value: 12345
    })

    expect(Comp.find(slector).first().text()).toBe('kr 12 345')
  })
  it('have to match currency with currency_position="after"', () => {
    const Comp = mount(
      <Component value={-value} currency currency_position="after" />
    )

    expect(Comp.find(slector).first().text()).toBe('-12 345 678,98 kr')

    expect(
      Comp.find(slector).first().instance().getAttribute('aria-label')
    ).toBe('-12 345 678,98 norske kroner')

    Comp.setProps({
      currency_display: 'code'
    })
    expect(Comp.find(slector).first().text()).toBe('-12 345 678,98 NOK')

    Comp.setProps({
      currency_position: 'before'
    })
    expect(Comp.find(slector).first().text()).toBe('NOK -12 345 678,98')
  })
  it('have to match currency under 100.000', () => {
    const Comp = mount(<Component value={-12345.95} currency />)

    expect(Comp.find(slector).first().text()).toBe('kr -12 345,95')

    expect(
      Comp.find(slector).first().instance().getAttribute('aria-label')
    ).toBe('-12345,95 norske kroner')
  })
  it('have to match currency with no decimals', () => {
    const Comp = mount(
      <Component value={-12345.99} currency decimals={0} />
    )

    expect(Comp.find(slector).first().text()).toBe('kr -12 345')

    expect(
      Comp.find(slector).first().instance().getAttribute('aria-label')
    ).toBe('-12345,00 norske kroner')
  })
  it('have to match phone number', () => {
    const Comp = mount(<Component phone>+47 99999999</Component>)
    expect(Comp.find(slector).first().text()).toBe('0047 99 99 99 99')
  })
  it('have to match bank account number', () => {
    const Comp = mount(<Component ban>20001234567</Component>)
    expect(Comp.find(slector).first().text()).toBe('2000 12 34567')

    // also check the formatting with one digit less
    Comp.setProps({
      children: null,
      value: 2000123456
    })
    expect(Comp.find(slector).first().text()).toBe('2000 12 3456')
  })
  it('have to match national identification number', () => {
    const Comp = mount(<Component nin>18089212345</Component>)
    expect(Comp.find(slector).first().text()).toBe('180892 12345')
    expect(
      Comp.find(slector).first().instance().getAttribute('aria-label')
    ).toBe('18 08 92 1 2 3 4 5')
  })
  it('have to match organization number', () => {
    const Comp = mount(
      <Component org suffix="MVA">
        123456789
      </Component>
    )
    expect(Comp.find(slector).first().text()).toBe('123 456 789 MVA')
    expect(
      Comp.find(slector).first().instance().getAttribute('aria-label')
    ).toBe('1 2 3 4 5 6 7 8 9 MVA')
  })
  it('have to handle prefix and suffix', () => {
    const Comp = mount(
      <Component prefix={<span>prefix</span>} suffix={<span>suffix</span>}>
        123456789.5
      </Component>
    )
    expect(Comp.find(slector).first().text()).toBe(
      'prefix 123 456 789,5 suffix'
    )
    expect(
      Comp.find(slector).first().instance().getAttribute('aria-label')
    ).toBe('prefix 123 456 789,5 suffix')
  })
  it('should validate with ARIA rules', async () => {
    const Comp = mount(<Component value={-value} currency />)
    expect(
      await axeComponent(Comp, {
        rules: {
          // because of the role="text", we disable this rule for now
          'aria-roles': { enabled: false }
        }
      })
    ).toHaveNoViolations()
  })
})

describe('Number cleanNumber', () => {
  it('should clean up and remove invalid suff arround numbers', () => {
    expect(cleanNumber(-12345.67)).toBe(-12345.67)
    expect(cleanNumber('prefix -12.345,678 suffix')).toBe('-12345.678')
    expect(cleanNumber('prefix -12 345,678 suffix')).toBe('-12345.678')
    expect(cleanNumber('prefix -12.345·678 suffix')).toBe('-12345.678')
    expect(cleanNumber("prefix -12.345'678 suffix")).toBe('-12345.678')
    expect(cleanNumber('prefix -12.345.678 suffix')).toBe('-12345678')
    expect(cleanNumber('prefix -1,234,567.891 suffix')).toBe(
      '-1234567.891'
    )
    expect(cleanNumber('prefix -1 234 567,891 suffix')).toBe(
      '-1234567.891'
    )
    expect(cleanNumber('prefix -1 234 567.891 suffix')).toBe(
      '-1234567.891'
    )
    expect(cleanNumber("prefix -1'234'567.891 suffix")).toBe(
      '-1234567.891'
    )
    expect(cleanNumber('prefix -1,234,567·891 suffix')).toBe(
      '-1234567.891'
    )
    expect(cleanNumber("prefix -1.234.567'891 suffix")).toBe(
      '-1234567.891'
    )
  })
})

describe('Number cleanDirtyNumber', () => {
  it('should clean up and remove invalid suff arround numbers', () => {
    expect(cleanDirtyNumber(-12345.67)).toBe('-12345.67')
    expect(cleanDirtyNumber('-12.345,67 suffix')).toBe('-12345.67')
    expect(cleanDirtyNumber('prefix -12.345,67')).toBe('-12345.67')
    expect(cleanDirtyNumber(' -12.345,67')).toBe('-12345.67')
    expect(cleanDirtyNumber('prefix -12 345,67 suffix')).toBe(false)
    expect(cleanDirtyNumber('prefix -12 345,67 $')).toBe(false)
    expect(cleanDirtyNumber('$ -12 345,67 suffix')).toBe(false)
    expect(cleanDirtyNumber(' -12 345,67 ')).toBe(false)
    expect(cleanDirtyNumber('  -12 345,67  ')).toBe(false)
    expect(cleanDirtyNumber('0047 ')).toBe('0047')
    expect(cleanDirtyNumber('prefix \n-12 345,67')).toBe(false)
    expect(cleanDirtyNumber('prefix')).toBe(false)
  })
})

describe('Number copy methods like', () => {
  it('copyNumber should make valid clipboard copy', async () => {
    copyNumber('1234.56')
    expect(await navigator.clipboard.readText()).toBe('1234.56')
  })
  it('copySelectedNumber make valid clipboard copy', async () => {
    copySelectedNumber()
    expect(await navigator.clipboard.readText()).toBe('1234.56')
  })
})

describe('Number scss', () => {
  it('have to match snapshot', () => {
    const scss = loadScss(require.resolve('../style/dnb-number.scss'))
    expect(scss).toMatchSnapshot()
  })
})
