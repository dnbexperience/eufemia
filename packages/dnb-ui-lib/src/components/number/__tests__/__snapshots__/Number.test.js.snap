// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Number component have to match default number snapshot 1`] = `
<Number
  ban={null}
  class={null}
  className={null}
  currency={null}
  currency_display={null}
  currency_position={null}
  decimals={null}
  element="span"
  link={null}
  locale="nb-NO"
  nin={null}
  options={null}
  org={null}
  phone={null}
  prefix={null}
  selectall={null}
  sufix={null}
  value={12345678.9876}
>
  <ForwardRef
    className="dnb-number"
    is="span"
    lang="nb-NO"
    onClick={[Function]}
    onCopy={[Function]}
    onTouchStart={[Function]}
    role="text"
  >
    <span
      className="dnb-number"
      lang="nb-NO"
      onClick={[Function]}
      onCopy={[Function]}
      onTouchStart={[Function]}
      role="text"
    >
      12 345 678,9876
    </span>
  </ForwardRef>
</Number>
`;

exports[`Number scss have to match snapshot 1`] = `
"/*
* DNB Number
*
*/
/**
 * This file is only used to make components independent
 * so that they can get imported individually, without the core styles
 *
 */
/*
 * Utilities
 */
/*
 * Scopes
 *
 */
/*
 * Document Reset
 *
 */
/*
 * Number Component
 *
 */
/*
 * Anchor mixins
 *
 */
.dnb-number {
  speak-as: numbers;
  color: inherit;
  cursor: text; }
  .dnb-number.dnb-anchor {
    cursor: pointer; }
  .dnb-number__fx {
    pointer-events: none;
    position: fixed;
    z-index: 3300;
    border-radius: 2px;
    background-color: var(--color-mint-green);
    color: var(--color-black);
    animation: number-fx-scale-out 750ms cubic-bezier(0.04, 0.65, 0.55, 0.95) 1 forwards; }
    .dnb-number__fx__selection {
      display: inline-block;
      width: 0;
      height: 0;
      font-size: 0;
      line-height: 0; }
  .dnb-number--selectall {
    user-select: all; }

@keyframes number-fx-scale-out {
  0% {
    opacity: 1;
    transform: scale(1); }
  90% {
    opacity: 0; }
  100% {
    opacity: 0;
    transform: scale(5); } }
"
`;
